File: maximum_flow/cpp/dinic.cpp
#include <fstream>
#include <iostream>
#include <limits>
#include <queue>
#include <sstream>
#include <vector>

struct Edge
{
    int to;
    int capacity;
    int flow;
    size_t reverse_edge_index;
};

class FlowNetwork
{
public:
    FlowNetwork(int vertices) : graph_(vertices), levels_(vertices), next_edge_tracking_(vertices) {}

    void AddingEdge(int from, int to, int capacity)
    {
        graph_[from].push_back({ to, capacity, 0, graph_[to].size() });
        graph_[to].push_back({ from, 0, 0, graph_[from].size() - 1 });
    }

    bool BFS(int source, int sink)
    {
        fill(levels_.begin(), levels_.end(), -1);
        levels_[source] = 0;

        std::queue<int> q;
        q.push(source);

        while (!q.empty())
        {
            int current = q.front();
            q.pop();

            for (const Edge& edge : graph_[current])
            {
                if (levels_[edge.to] == -1 && edge.capacity > edge.flow)
                {
                    levels_[edge.to] = levels_[current] + 1;
                    q.push(edge.to);
                }
            }
        }

        return levels_[sink] != -1;
    }

    int DFS(int current, int sink, int minimum_capacity)
    {
        if (current == sink)
        {
            return minimum_capacity;
        }

        for (int& i = next_edge_tracking_[current]; i < graph_[current].size(); ++i)
        {
            Edge& edge = graph_[current][i];
            if (levels_[edge.to] == levels_[current] + 1 && edge.capacity > edge.flow)
            {
                int flow = DFS(edge.to, sink, std::min(minimum_capacity, edge.capacity - edge.flow));
                if (flow > 0)
                {
                    edge.flow += flow;
                    graph_[edge.to][edge.reverse_edge_index].flow -= flow;
                    return flow;
                }
            }
        }

        return 0;
    }

    int Dinic(int source, int sink)
    {
        int maximum_flow = 0;

        while (BFS(source, sink))
        {
            fill(next_edge_tracking_.begin(), next_edge_tracking_.end(), 0);
            while (int flow = DFS(source, sink, std::numeric_limits<int>::max()))
            {
                maximum_flow += flow;
            }
        }

        return maximum_flow;
    }

    std::vector<std::vector<Edge>> graph_;

private:
    std::vector<int> levels_;
    std::vector<int> next_edge_tracking_;
};

int main()
{
    std::ios::sync_with_stdio(false);
    std::cin.tie(NULL);
    std::ifstream input_file("maximum_flow_input.txt");

    if (!input_file)
    {
        std::cerr << "Failed to open the file." << std::endl;
        return 1;
    }

    int vertices, edges, source, sink;
    input_file >> vertices >> edges;

    std::cout << "Enter the source node: ";
    std::cin >> source;
    std::cout << "Enter the sink node: ";
    std::cin >> sink;

    FlowNetwork network(vertices);

    for (int i = 0; i < vertices; ++i)
    {
        network.graph_.push_back({});
    }

    for (int i = 0; i < edges; ++i)
    {
        int from, to, capacity;
        input_file >> from >> to >> capacity;
        network.AddingEdge(from, to, capacity);
    }

    input_file.close();

    int maximum_flow = network.Dinic(source, sink);
    std::cout << "Maximum Flow: " << maximum_flow << std::endl;

    return 0;
}

File: maximum_flow/python/dinic.py
from collections import deque

class Edge:
    def __init__(self, to, capacity, flow, reverse_edge_index):
        self.to = to
        self.capacity = capacity
        self.flow = flow
        self.reverse_edge_index = reverse_edge_index

class FlowNetwork:
    def __init__(self, vertices):
        self.graph_ = [[] for _ in range(vertices)]
        self.levels_ = [-1] * vertices
        self.next_edge_tracking_ = [0] * vertices

    def add_edge(self, from_, to, capacity):
        self.graph_[from_].append(Edge(to, capacity, 0, len(self.graph_[to])))
        self.graph_[to].append(Edge(from_, 0, 0, len(self.graph_[from_]) - 1))

    def bfs(self, source, sink):
        self.levels_ = [-1] * len(self.levels_)
        self.levels_[source] = 0

        q = deque([source])

        while q:
            current = q.popleft()
            for edge in self.graph_[current]:
                if self.levels_[edge.to] == -1 and edge.capacity > edge.flow:
                    self.levels_[edge.to] = self.levels_[current] + 1
                    q.append(edge.to)

        return self.levels_[sink] != -1

    def dfs(self, current, sink, minimum_capacity):
        if current == sink:
            return minimum_capacity

        while self.next_edge_tracking_[current] < len(self.graph_[current]):
            edge = self.graph_[current][self.next_edge_tracking_[current]]
            if self.levels_[edge.to] == self.levels_[current] + 1 and edge.capacity > edge.flow:
                flow = self.dfs(edge.to, sink, min(minimum_capacity, edge.capacity - edge.flow))
                if flow > 0:
                    edge.flow += flow
                    self.graph_[edge.to][edge.reverse_edge_index].flow -= flow
                    return flow

            self.next_edge_tracking_[current] += 1

        return 0

    def dinic(self, source, sink):
        maximum_flow = 0

        while self.bfs(source, sink):
            self.next_edge_tracking_ = [0] * len(self.next_edge_tracking_)
            while flow := self.dfs(source, sink, float('inf')):
                maximum_flow += flow

        return maximum_flow

if __name__ == '__main__':
    with open("maximum_flow_input.txt", 'r') as input_file:
        vertices, edges = map(int, input_file.readline().split())
        network = FlowNetwork(vertices)

        for _ in range(edges):
            from_, to, capacity = map(int, input_file.readline().split())
            network.add_edge(from_, to, capacity)

    source = int(input("Enter the source node: "))
    sink = int(input("Enter the sink node: "))

    if (source == sink):
        print("Source and sink are the same... No flow!")
        exit()
    else:
        maximum_flow = network.dinic(source, sink)
        print("Maximum Flow:", maximum_flow)
        